Sources/CNIOLinux/liburing_shims.c

```diff
@@ -0,0 +1,545 @@
+        fprintf(stderr, "WARNING: Failed to load liburing.so, using epoll() instead. [%s] [%p]\n", err?err:"Unknown reason", dl_handle);
+        return -1;
+    }
+    
+     // try to resolve all symbols we need, macro will fail with -1 if unsuccessful
+    _DL_RESOLVE(io_uring_get_probe_ring);
+    _DL_RESOLVE(io_uring_get_probe);
+    _DL_RESOLVE(io_uring_free_probe);
+    _DL_RESOLVE(io_uring_queue_init_params);
+    _DL_RESOLVE(io_uring_queue_init);
+    _DL_RESOLVE(io_uring_queue_mmap);
+    _DL_RESOLVE(io_uring_ring_dontfork);
+    _DL_RESOLVE(io_uring_queue_exit);
+    _DL_RESOLVE(io_uring_peek_batch_cqe);
+    _DL_RESOLVE(io_uring_wait_cqes);
+    _DL_RESOLVE(io_uring_wait_cqe_timeout);
+    _DL_RESOLVE(io_uring_submit);
+    _DL_RESOLVE(io_uring_submit_and_wait);
+    _DL_RESOLVE(io_uring_get_sqe);
+    _DL_RESOLVE(io_uring_register_buffers);
+    _DL_RESOLVE(io_uring_unregister_buffers);
+    _DL_RESOLVE(io_uring_register_files);
+    _DL_RESOLVE(io_uring_unregister_files);
+    _DL_RESOLVE(io_uring_register_files_update);
+    _DL_RESOLVE(io_uring_register_eventfd);
+    _DL_RESOLVE(io_uring_register_eventfd_async);
+    _DL_RESOLVE(io_uring_unregister_eventfd);
+    _DL_RESOLVE(io_uring_register_probe);
+    _DL_RESOLVE(io_uring_register_personality);
+    _DL_RESOLVE(io_uring_unregister_personality);
+    _DL_RESOLVE(__io_uring_sqring_wait);
+    _DL_RESOLVE(__io_uring_get_cqe);
+        
+    // Finally after resolving, probe actual capabilities
+    if (_check_capabilities() != 0) {
+        (void) dlclose(dl_handle); // we don't care about errors, nothing we can do anyway
+        return -1;
+    }
+
+    return 0;
+}
+
+// And the wrappers, should never be called unless we've done CNIOLinux_io_uring_load once first.
+
+pthread_mutex_t global_ring_mutex = PTHREAD_MUTEX_INITIALIZER;
+static struct io_uring global_ring; // shared small ring to be able to reuse SQPOLL kernel thread
+
+// FIXME: We need to look at cpu affinity and preferably bind the SQPOLL thread to a specific core
+// and support cgroup / cpu sets on Linux to isolate it - then we should set IORING_SETUP_SQ_AFF here.
+
+int CNIOLinux_io_uring_queue_init(unsigned entries, struct io_uring *ring,
+    unsigned flags)
+{
+    if (flags & IORING_SETUP_SQPOLL)
+    {
+        pthread_mutex_lock(&global_ring_mutex);
+        if (!global_ring.ring_fd)
+        {
+            // setup a small global ring whose fd we can reference to use shared kernel sqpoll thread for all rings
+            // we don't particulariy care if we fail, then no SQPOLL shared ring
+            (void) liburing_functions.io_uring_queue_init(4, &global_ring, IORING_SETUP_SQPOLL); // should support IORING_SETUP_SQ_AFF
+        }
+        pthread_mutex_unlock(&global_ring_mutex);
+
+        if (global_ring.ring_fd) // use shared kernel thread if it exists, otherwise fallthrough to normal setup
+        {
+            struct io_uring_params params;
+            memset(&params, 0, sizeof(params));
```

Comment:
this isn't necessary in C

    struct io_uring_params params = { 0 };

is nicer and less error prone.